zarr/tests/test_indexing.py

```diff
@@ -0,0 +1,877 @@
+       replace_ellipsis((Ellipsis, slice(None), slice(None)), (100, 100)))
+    eq((slice(None), slice(None)),
+       replace_ellipsis((slice(None), slice(None), Ellipsis), (100, 100)))
+
+
+def _test_get_orthogonal_selection_1d_common(a, z, ix):
+    expect = a[ix]
+    actual = z.get_orthogonal_selection(ix)
+    assert_array_equal(expect, actual)
+    actual = z.oindex[ix]
+    assert_array_equal(expect, actual)
+    # for 1d arrays, also available via __getitem__
+    actual = z[ix]
+    assert_array_equal(expect, actual)
+
+
+# noinspection PyStatementEffect
+def test_get_orthogonal_selection_1d_bool():
+
+    # setup
+    a = np.arange(1050, dtype=int)
+    z = zarr.create(shape=a.shape, chunks=100, dtype=a.dtype)
+    z[:] = a
+
+    np.random.seed(42)
+    # test with different degrees of sparseness
+    for p in 0.5, 0.1, 0.01:
+        ix = np.random.binomial(1, p, size=a.shape[0]).astype(bool)
+        _test_get_orthogonal_selection_1d_common(a, z, ix)
+
+    # test errors
+    with assert_raises(IndexError):
+        z.oindex[np.zeros(50, dtype=bool)]  # too short
+    with assert_raises(IndexError):
+        z.oindex[np.zeros(2000, dtype=bool)]  # too long
+    with assert_raises(IndexError):
+        z.oindex[[[True, False], [False, True]]]  # too many dimensions
+
+
+# noinspection PyStatementEffect
+def test_get_orthogonal_selection_1d_int():
+
+    # setup
+    a = np.arange(1050, dtype=int)
+    z = zarr.create(shape=a.shape, chunks=100, dtype=a.dtype)
+    z[:] = a
+
+    np.random.seed(42)
+    # test with different degrees of sparseness
+    for p in 2, 0.5, 0.1, 0.01:
+        ix = np.random.choice(a.shape[0], size=int(a.shape[0] * p), replace=True)
+        _test_get_orthogonal_selection_1d_common(a, z, ix)
+        ix.sort()
+        _test_get_orthogonal_selection_1d_common(a, z, ix)
+
+    # test wraparound
+    ix = [0, 3, 10, -23, -12, -1]
+    expect = a[ix]
+    actual = z.oindex[ix]
+    assert_array_equal(expect, actual)
+
+    # explicit test not sorted
+    ix = [3, 105, 23, 127]  # not monotonically increasing
+    expect = a[ix]
+    actual = z.oindex[ix]
+    assert_array_equal(expect, actual)
+
+    # test errors
+    with assert_raises(IndexError):
+        ix = [a.shape[0] + 1]  # out of bounds
+        z.oindex[ix]
+    with assert_raises(IndexError):
+        ix = [-(a.shape[0] + 1)]  # out of bounds
+        z.oindex[ix]
+    with assert_raises(IndexError):
+        ix = [[2, 4], [6, 8]]  # too many dimensions
+        z.oindex[ix]
+
+
+def test_get_orthogonal_selection_1d_slice_with_step():
+
+    # setup
+    a = np.arange(1050, dtype=int)
+    z = zarr.create(shape=a.shape, chunks=100, dtype=a.dtype)
+    z[:] = a
+
+    selections = [
+        slice(0, 1050),
+        slice(0, 1050, 1),
+        slice(0, 1050, 10),
+        slice(0, 1050, 100),
+        slice(0, 1050, 1000),
+        slice(50, 150, 1),
+        slice(50, 150, 10),
+        slice(50, 150, 100),
+    ]
```

Comment:
What about negative steps? At the least, those should give an appropriate error.

Reply:
Negative steps are supported, I've added tests to confirm.